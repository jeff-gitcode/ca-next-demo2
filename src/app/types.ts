export const TYPES = {
  TodoUseCase: Symbol.for('TodoUseCase'),
  TodoService: Symbol.for('TodoService'),
  TodoRepository: Symbol.for('TodoRepository'),
};
