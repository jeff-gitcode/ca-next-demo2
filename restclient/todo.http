@host=http://127.0.0.1:8090


### GET todo records
GET {{host}}/api/collections/todo/records?page=1&perPage=10 HTTP/1.1
content-type: application/json

### GET todo record
GET {{host}}/api/collections/todo/records/xucm9od3kwqq7zq HTTP/1.1
content-type: application/json

### CREATE todo record
POST {{host}}/api/collections/todo/records HTTP/1.1
accept: text/plain
content-type: application/json

{
  "title": "Admin1",
  "content": ""
}

### Update todo record
PATCH {{host}}/api/collections/todo/records/xucm9od3kwqq7zq HTTP/1.1
accept: text/plain
content-type: application/json

{
  "title": "Admin updated",
  "content": "content1"
}

### DELETE todo record
DELETE  {{host}}/api/collections/todo/records/szfhv770n23b80j HTTP/1.1
accept: text/plain
content-type: application/json


### GET todo records from nextjs
GET http://127.0.0.1:3000/api/todos HTTP/1.1
content-type: application/json

### GET todo record
GET http://127.0.0.1:3000/api/todos/xucm9od3kwqq7zq HTTP/1.1
content-type: application/json

### CREATE todo record
POST http://127.0.0.1:3000/api/todos HTTP/1.1
accept: text/plain
content-type: application/json

{
    "title": "Admin"
}

### UPDATE todo record
PATCH http://127.0.0.1:3000/api/todos/xucm9od3kwqq7zq HTTP/1.1
accept: text/plain
content-type: application/json

{
    "title": "Admin UPDATED"
}

### DELETE todo record
DELETE  http://127.0.0.1:3000/api/todos/3cyet2vmpwj7h2a HTTP/1.1
accept: text/plain
content-type: application/json

### LoginUserFunction (User)
POST {{host}}/api/LoginUserFunction HTTP/1.1
accept: text/plain
content-type: application/json

{
  "name": "Jeff",
  "role": "User",
  "email": "user@email.com",
  "password": "password"
}

### LoginUserFunction (Admin)
POST {{host}}/api/LoginUserFunction HTTP/1.1
accept: text/plain
content-type: application/json

{
  "name": "Admin",
  "role": "Admin",
  "email": "admin@email.com",
  "password": "password"
}

### SearchUserFunction
POST {{host}}/api/SearchUserFunction HTTP/1.1
Authorization: Bearer {{token}}
accept: text/plain
content-type: application/json

{
  "name": "John",
  "page": 1,
  "size": 1
}
