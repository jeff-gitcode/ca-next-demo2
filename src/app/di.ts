import { Container } from "inversify";

import { infrastrcutrueModule } from "./infrastrcutrue/di";
import { applicationModule } from "./application/di";

const AppplicationContainer = new Container();

AppplicationContainer.load(infrastrcutrueModule);
AppplicationContainer.load(applicationModule);
