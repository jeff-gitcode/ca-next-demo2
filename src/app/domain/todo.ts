export type Todo = {
  readonly id: string;
  readonly title: string;
};
