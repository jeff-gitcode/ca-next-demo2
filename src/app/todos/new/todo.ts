import z from "zod";

export const todoSchema = z.object({
    id: z.string(),
    title: z.string().min(1),
});

export type FormValue = z.infer<typeof todoSchema>;