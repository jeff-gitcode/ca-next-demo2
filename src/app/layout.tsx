'use client';

import { Inter } from 'next/font/google';
import './globals.css';
import Link from 'next/link';

import { ApplicationContainer } from './di';
import { useState } from 'react';
import { ApplicationProvider } from './presentation/applicationProvider';

const inter = Inter({ subsets: ['latin'] });

// export const metadata: Metadata = {
//   title: "Create Next App",
//   description: "Generated by create next app",
// };

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <ApplicationProvider container={ApplicationContainer}>
          <div>
            <header> </header>
            <nav>
              <Link href="/presentation/todos">Todos</Link> |
              <Link href="/">About</Link>
            </nav>
          </div>
          {children}
        </ApplicationProvider>
      </body>
    </html>
  );
}
