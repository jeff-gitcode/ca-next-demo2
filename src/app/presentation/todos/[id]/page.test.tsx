import { fireEvent, prettyDOM, render, screen, waitFor } from "@testing-library/react";
import { TodoUseCase } from "@/app/application/hooks/use.todo.controller";
import userEvent from "@testing-library/user-event";

import { FormValue, todoSchema } from "@/app/application/validator/todo";
import { useRouter } from "next/navigation";
import * as ApplicationContext from "@/app/application/hooks/app.context";
import { act } from "react";
import TodoPage from "./page";
import { NavigateOptions } from "next/dist/shared/lib/app-router-context.shared-runtime";

const mockPush = jest.fn();
jest.mock("next/navigation", () => ({
    useRouter() {
        return {
            query: {
                id: "1",
            },
            push: mockPush,
        };
    },
}));

jest.mock("@/app/application/hooks/app.context", () => {
    const original: typeof ApplicationContext = jest.requireActual("@/app/application/hooks/app.context");

    return ({
        ...original,
        useAppContext: jest.fn(),
    });
});
const mockUseAppContext = ApplicationContext.useAppContext as jest.Mock;

function setup(ui: React.ReactNode) {
    return {
        user: userEvent.setup(),
        ...render(ui),
    };
}

describe("TodoPage", () => {
    let mockTodoUseCase: jest.Mock;
    let mockUpdateTodoUseCase: jest.Mock;

    beforeEach(() => {
        jest.setTimeout(10000);
        jest.clearAllMocks();

        mockTodoUseCase = jest.fn();
        mockUpdateTodoUseCase = jest.fn();

        const contextValues = {
            todoUseCase: mockTodoUseCase,
            updateTodoUseCase: mockUpdateTodoUseCase,
        };

        mockUseAppContext.mockReturnValue(contextValues);
    });

    test("should render", () => {
        // Arrange
        const data = {
            id: "1",
            title: "test title",
            description: "test description",
            status: "test status",
            created_at: "test created_at",
            updated_at: "test updated_at",
        };

        const isLoading = true;
        const error = undefined;
        const updateData = undefined;

        mockTodoUseCase.mockReturnValue({ data, isLoading, error });
        mockUpdateTodoUseCase.mockReturnValue({ updateData });

        // Act
        const { container } = render(<TodoPage params={{ id: "1" }} />);

        // Assert
        screen.debug();

        expect(screen.getByPlaceholderText('ID')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Title')).toBeInTheDocument();
        expect(screen.getByRole('button', { name: 'Submit' })).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test("should render loading", () => {
        // Arrange
        const data = undefined;
        const isLoading = true;
        const error = undefined;
        const updateData = undefined;

        mockTodoUseCase.mockReturnValue({ data, isLoading, error });
        mockUpdateTodoUseCase.mockReturnValue({ updateData });

        // Act
        render(<TodoPage params={{ id: "1" }} />);

        // Assert
        expect(screen.getByText('loading...')).toBeInTheDocument();
    });

    test("should render error", () => {
        // Arrange
        const data = undefined;
        const isLoading = false;
        const error = new Error("test error");
        const updateData = undefined;

        mockTodoUseCase.mockReturnValue({ data, isLoading, error });
        mockUpdateTodoUseCase.mockReturnValue({ updateData });

        // Act
        render(<TodoPage params={{ id: "1" }} />);

        // Assert
        expect(screen.getByText('failed to load')).toBeInTheDocument();
    });

    test("should update todo", async () => {
        // Arrange
        const data = {
            id: "1",
            title: "test title",
            description: "test description",
            status: "test status",
            created_at: "test created_at",
            updated_at: "test updated_at",
        };

        const isLoading = false;
        const error = undefined;
        const updateData = undefined;
        const mockUpdateTodo = jest.fn();
        const updateTodo = mockUpdateTodo;
        const isUpdating = false;

        mockTodoUseCase.mockReturnValue({ data, isLoading, error });
        mockUpdateTodoUseCase.mockReturnValue({ updateData, updateTodo, isUpdating });

        // Act
        // const { user } = setup(<TodoPage params={{ id: "1" }} />);
        render(<TodoPage params={{ id: "1" }} />);
        const title = screen.getByPlaceholderText('Title');
        await fireEvent.input(title, { target: { value: 'updated title' } });
        screen.debug();

        // Assert
        const confirmButton = screen.getByTestId('submit');
        await waitFor(() => {
            fireEvent.click(confirmButton);
            // console.log(confirmButton);
            console.log(prettyDOM(confirmButton));
        });

        // await waitFor(() => {
        expect(mockUpdateTodo).toHaveBeenCalledWith({ requestBody: { id: "1", title: "updated title" }, queryParams: { id: "1" } });
        expect(mockPush).toHaveBeenCalledWith(`/presentation/todos`);
        // });
    }, 30000);
});

