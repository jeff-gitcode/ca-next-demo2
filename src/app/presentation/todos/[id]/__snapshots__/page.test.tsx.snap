// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPage should render 1`] = `
<div>
  <form
    class="flex flex-col items-center space-y-2"
  >
    <input
      class="text-black flex rounded-lg px-1 py-1 bg-grey-100 peer block"
      disabled=""
      name="id"
      placeholder="ID"
    />
    <input
      class="text-black flex rounded-lg px-1 py-1"
      name="title"
      placeholder="Title"
    />
    <button
      class="text-white bg-blue-700 font-medium rounded-lg text-sm px-5 py-2.5"
      type="submit"
    >
      Confirm
    </button>
  </form>
</div>
`;
